Graph coloring ZKP

Syndamia opened the pull request:
A Zero-Knowledge Program, which takes as input:

1. public input: an undirected unweighted graph
2. public input: number of colors and
3. private input: coloring of the graph with exactly this many colors

and checks whether the coloring is correct.

Syndamia committed:
feat(graph-coloring): Add implementation of a graph coloring Noir program

Syndamia committed:
feat(graph-coloring): Create adjacency matrix from prover

Syndamia committed:
feat(requirements.txt): Add pip requirements.txt

smanilov commented on graph-coloring/generate-graph/__main__.py:
```diff
@@ -0,0 +1,21 @@
+def to_toml_arr(x):
+    return list(map(list, x))
+
+n = argv[2] if len(argv) > 2 else 100
```
`argv[2]` will be str; should be `int(argv[2])`

smanilov commented on graph-coloring/generate-graph/__main__.py:
```diff
@@ -0,0 +1,21 @@
+import scipy
+argv = sys.argv
+
+sys.stdout = open(argv[1], "wt")
```
Perhaps there should be a default value for argv[1] as well, e.g. "Prover.toml".

smanilov commented on graph-coloring/Makefile:
```diff
@@ -0,0 +1,7 @@
+all: generate-prover execute
+
+generate-prover:
+	python ./generate-graph
```
This does not work now, because the first argument is mandatory. If you add a default value it should work.

smanilov commented on graph-coloring/generate-graph/__main__.py:
```diff
@@ -0,0 +1,21 @@
+    return list(map(list, x))
+
+n = argv[2] if len(argv) > 2 else 100
+m = argv[3] if len(argv) > 3 else 1.5 * n
```
1.5 * n is of type float; wrap in int() as well, otherwise this throws an error later

smanilov commented on graph-coloring/src/main.nr:
```diff
@@ -0,0 +1,28 @@
+mod graph;
+use graph::{ Vertex, Graph };
+
+global VERTECIES: u32 = 100;
```
100 is too much. Let's have 10 for now.

smanilov commented on graph-coloring/src/graph.nr:
```diff
@@ -0,0 +1,63 @@
+
+		for _ in 0..Vertecies {
+			// If we have disconnected components, then we'll need to add start
+			// vertecies for each component
```
typo: "vertecies" should be "vertices"

smanilov reviewed (commented):
A few minor comments.

In addition to those, please, move the python code in a directory called `greedy-gen` and the noir code in a directory called `typeful-zk` (to indicate that you went for a fancy solution; could be `fancy`, if you prefer).

My idea is that we also add `bruteforce-gen` and `simple-zk` as a follow-up.

Syndamia committed:
feat: Rename generate-graph to greedy-gen and move Noir program to structful-zk

Syndamia committed:
fix(greedy-gen): Add default for argv[1], add casts to int

Syndamia committed:
fix(greedy-gen): Reduce default vertecies to 10

Syndamia committed:
fix(structful-zk): Reduce vertecies to 10

Syndamia committed:
fix(structful-zk): Rename vertecies to vertices

Syndamia committed:
fix(structful-zk): Update Makefile to be useable

Syndamia committed:
feat(graph-coloring): Add a very basic ZKP to complement structful-zk

Syndamia committed:
feat(graph-coloring/greedy-gen): Rework python module

smanilov committed:
feat(graph-coloring): Add README.md

smanilov reviewed (approved):
Excellent!

I added a README.md in a new commit.

Feel free to edit before merging if you feel like it's necessary.

Otherwise, you can just merge.

Syndamia committed:
feat(graph-coloring/Makefile): Add ability to choose vertices and/or edges

Have to set variables V and E (the latter of which is optional)
Those variables may be exported or passed like `make V=20 E=40`

Syndamia committed:
feat(graph-coloring): Update README with a lot more detail
